srcjs/shinyapp.js

```diff
@@ -631,6 +631,38 @@ var ShinyApp = function() {
     window.location.reload();
   });
 
+  addMessageHandler('shiny-insert-ui', function (message) {
+      var targets = $(message.selector);
+      if (targets.length === 0) {
+        // render the HTML and deps to a null target, so
+        // the side-effect of rendering the deps, singletons,
+        // and <head> still occur
+        exports.renderHtml($([]), message.content.html, message.content.deps);
+      } else {
+        targets.each(function (i, target) {
+          var container = document.createElement(message.container);
+          insertAdjacentElement(message.where, target, container);
+          exports.renderContent(container, message.content);
+          $(container).trigger('shown');
+          return message.multiple;
+        });
+      }
+    });
+
+  addMessageHandler('shiny-remove-ui', function (message) {
+    var els = $(message.selector);
+    els.each(function (i, el) {
+      exports.unbindAll(el, true);
+      $(el).trigger('hide');
```

Comment:
Hidden state absolutely needs to be updated for a plot you're removing, otherwise the server will continue to render it even though it will never be shown. However, sendOutputHiddenState is (very sensibly) called by unbindOutputs--I had forgotten that. So we're covered there.

But we still do need to at least call sendImageSize from shiny-remove-ui. (This doesn't need to be called from htmlOutputBinding because that code path calls `renderContent`>`bindAll`>`bindOutputs`>`sendImageSize`.) This matters for cases where the removal of an element could cause a different element to change size--for example, fillRow, fillCol, flexdashboard, or even just the height of the page shrinking to where the scrollbar is no longer needed.